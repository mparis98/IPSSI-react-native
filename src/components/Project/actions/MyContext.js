import React, {useState, useEffect} from 'react';
import {Alert, AsyncStorage} from 'react-native';

// this is the equivalent to the createStore method of Redux
// https://redux.js.org/api/createstore

export const MyContext = React.createContext(null);

export const FindCoordinatesProvider = props => {
    const [location, setLocation] = useState(null);

    const getData = (latitude,longitude,distance)=> {
        return `https://opendata.paris.fr/api/records/1.0/search/?dataset=velib-disponibilite-en-temps-reel&rows=200&geofilter.distance=${latitude}%2C${longitude}%2C${distance}`;
    };

    const get = async (url, networkPriority = false) => {
        const key = encodeURIComponent(url);
        const value = await AsyncStorage.getItem(key)

        try {
            if (value !== null && !networkPriority) {
                return {data: JSON.parse(value), from: "Cache"};
            } else {
                const response = await fetch(url);
                const json = await response.json();
                await AsyncStorage.setItem(key, JSON.stringify(json));

                return {data: json, from: 'API'}
            }

        } catch (error) {
            return {
                data: JSON.parse(value),
                from: 'CACHE',
                error: error.message
            }
        }
    };

    const [data, setData] = useState(null);
    const [loading, setLoading] = useState(null);

    const fetchData = (networkPriority,position) => {
        setLoading(true);
            get(getData(position.coords.latitude,position.coords.longitude,1000), networkPriority).then(data => {
                setData(data);
                setLoading(false);
            });
    };

    useEffect(() => {
        navigator.geolocation.getCurrentPosition(
            position => {
                setLocation({
                    latitude:position.coords.latitude,
                    longitude:position.coords.longitude,
                    distance: 1000
                });
                fetchData(false,position);
                fetchData(true,position);

            },
            error => Alert.alert(error.message),
            {enableHighAccuracy: true, timeout: 20000, maximumAge: 1000}
        );
    }, [])


    return <MyContext.Provider value={{ location, setLocation, data, setData, loading, setLoading }} {...props} />;
};

