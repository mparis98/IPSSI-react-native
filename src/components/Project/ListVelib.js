import {AsyncStorage, View, Text, StyleSheet, FlatList} from 'react-native';
import React, {useState, useEffect} from 'react';
import {MyContext} from "./actions/MyContext";

const Velib = () => {
    return (
        <MyContext.Consumer>
            {context => (
                <View style={styles.container}>
                    {!context.data && <Text>Chargement...</Text>}
                    {context.data && (
                        <>
                            {console.log(context.data)}
                            <View style={styles.view}>
                            {context.data.data && <Text>{context.data.data.nhits} résultats</Text>}
                            <Text>From {context.data.from}</Text>
                            {context.data.error && <Text>Error: {JSON.stringify(context.data.error)}</Text>}
                            </View>
                            <FlatList data={context.data.data.records} renderItem={({ item }) =>
                                <View style={styles.item}>
                                <Text>{item.fields.station_name}</Text>
                                    <Text>{item.fields.station_state}</Text>
                                    <Text>Vélo disponible : {item.fields.nbbike}</Text>
                                    <Text>Vélo électrique disponible : {item.fields.nbebike}</Text>
                                </View>
                            }
                                      keyExtractor={item => item.fields.station_code} />
                        </>
                    )}
                </View>
            )}
        </MyContext.Consumer>
    )

};

const styles = {
    container: {
        flex: 1,
        marginTop: 10,
    },
    item: {
        backgroundColor: '#bcffac',
        padding: 20,
        marginVertical: 8,
        marginHorizontal: 16,
    },
    view: {
        padding: 20,
        marginVertical: 8,
        marginHorizontal: 16,
        textAlign: 'center',
        justifyContent: 'center'
    },
};


export default Velib;
