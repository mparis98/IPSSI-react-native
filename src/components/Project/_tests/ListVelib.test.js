import React from 'react';
import ListVelib from '../ListVelib';
import renderer from 'react-test-renderer';
import {MyContext} from "../actions/MyContext";


it('renders correctly', () => {
    const data = {
        data: {
            data: {
                nhits: 1,
                records: [{
                    fields: {
                        nbbike: 10,
                        nbebike: 4,
                        station_code: 11111,
                        station_name: 'Test Snapshot',
                        station_state: 'Operative'
                    }
                }]
            }
        }
    };

    const tree = renderer.create(<MyContext.Provider value={data}><ListVelib/></MyContext.Provider>).toJSON();
    expect(tree).toMatchSnapshot();
});
