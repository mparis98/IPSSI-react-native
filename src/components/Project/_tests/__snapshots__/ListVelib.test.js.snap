// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "marginTop": 10,
    }
  }
>
  <View
    style={
      Object {
        "justifyContent": "center",
        "marginHorizontal": 16,
        "marginVertical": 8,
        "padding": 20,
        "textAlign": "center",
      }
    }
  >
    <Text>
      1
       résultats
    </Text>
    <Text>
      From 
    </Text>
  </View>
  <RCTScrollView
    data={
      Array [
        Object {
          "fields": Object {
            "nbbike": 10,
            "nbebike": 4,
            "station_code": 11111,
            "station_name": "Test Snapshot",
            "station_state": "Operative",
          },
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#bcffac",
              "marginHorizontal": 16,
              "marginVertical": 8,
              "padding": 20,
            }
          }
        >
          <Text>
            Test Snapshot
          </Text>
          <Text>
            Operative
          </Text>
          <Text>
            Vélo disponible : 
            10
          </Text>
          <Text>
            Vélo électrique disponible : 
            4
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
